### PR TITLE
Feature/dp 831 international identifiers add update organisation

### PR DESCRIPTION
**Pull request by shilpigoel-goaco**

### CODE REVIEW

File: Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    //    var schemesInRegistries = InternationalIdentifiers.Select(x => x.Scheme).ToHashSet();
+    //    var existingInRegistries = ExistingInternationalIdentifiers.Where(scheme => schemesInRegistries.Contains(scheme) || scheme == Country + "-Other").ToList();
+
+    //    if (schemesInRegistries.Any() && existingInRegistries.Count == 1)
+    //    {
+    //        HasIdentifierToShow = true;
+    //    }
+
+    //    return true;
+    //}
```
Comment (andymantell) [inline Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs:132]: Probably remove this unless it's coming back imminently

**Review by andymantell — CHANGES_REQUESTED** [review]

File: Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs (hunk 1 above)
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs:132]: done

**Review by andymantell — APPROVED** [review]

**Review by dbgoaco — APPROVED** [review]

File: Frontend/CO.CDP.OrganisationApp/CO.CDP.OrganisationApp.csproj [hunk 2]
```diff
@@ -29,6 +29,6 @@
       <ProjectReference Include="..\..\Libraries\CO.CDP.Mvc.Validation\CO.CDP.Mvc.Validation.csproj" />
       <ProjectReference Include="..\..\Libraries\CO.CDP.Configuration\CO.CDP.Configuration.csproj" />
       <ProjectReference Include="..\..\Services\CO.CDP.Localization\CO.CDP.Localization.csproj" />
-    </ItemGroup>
+    </ItemGroup>
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp/CO.CDP.OrganisationApp.csproj:32]: Unnecessary change.

**Review by jakzal — CHANGES_REQUESTED** [review]

File: Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            ApiExceptionMapper.MapApiExceptions(aex, ModelState);
+            return Page();
+        }
+        catch (ApiException ex) when (ex.StatusCode == 404)
+        {
+            return Redirect("/page-not-found");
+        }
+    }
+
+    //private async Task<bool> ValidateAndSetExistingIdentifiers()
```
[the remark this answers is not in the document]
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs]: sorted.

File: Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    private async Task IdentifierRegistries()
+    {
+        try
+        {
+            InternationalIdentifiers = await pponClient.GetIdentifierRegistriesAsync(Country);
+        }
+        catch (ApiException ex) when (ex.StatusCode == 404)
+        {
+            // Show other
```
[the remark this answers is not in the document]
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs]: removed.

File: Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private async Task<ICollection<IdentifierRegistries>> PopulateIdentifierRegistries()
+    {
+        try
+        {
+            return await pponClient.GetIdentifierRegistriesAsync(Country);
+        }
+        catch (ApiException ex) when (ex.StatusCode == 404)
+        {
+            return new List<IdentifierRegistries>();
+            // Show other
```
[the remark this answers is not in the document]
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Pages/Organisation/OrganisationInternationalIdentification.cshtml.cs]: Actually pponClient.GetIdentifierRegistriesAsync throws 404 if no identifier found so I will have to catch the exception and Other option should be displayed.
But I have removed the commented line.

File: Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs [hunk 6]
```diff
@@ -38,7 +75,7 @@ public class RequiredIfContainsAttribute(string dependentProperty, string contai
         // Check if value is null or empty when condition is met
         if (conditionMet && (value == null || (value is List<string> str && str.Count == 0)))
```
[the remark this answers is not in the document]
Comment (shilpigoel-goaco) [reply Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs:76]: It is aligned with other attribute validators. if I change conditionMet variable in RequiredIfHasValueAttribute then it will needs to be changed in others also. 
Please suggest?

[the remark this answers is not in the document]
Comment (dbgoaco) [reply Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs:76]: My view on PRs is I want the code to read to me. conditionMet tells me nothing, I can see it's used as part of a condition statement but what does this condition represent? Also the comment on line 75 should be removed, it doesn't tell us anything more than the code does.

File: Frontend/CO.CDP.OrganisationApp/CO.CDP.OrganisationApp.csproj (hunk 2 above)
Comment (jakzal) [reply Frontend/CO.CDP.OrganisationApp/CO.CDP.OrganisationApp.csproj:32]: If the trailing spaces were added accidentaly, they should be removed to keep the PR clean and changes relevant. Next someone's in this file and their IDE reformats it, the change will be reverted and again cause an unrelated change on the PR.

File: Frontend/CO.CDP.OrganisationApp/Constants/Country.cs [hunk 7]
```diff
@@ -208,47 +170,40 @@ public static class Country
...
             {"VE", "Venezuela"},
             {"VN", "Vietnam"},
-            {"VG", "Virgin Islands, British"},
-            {"VI", "Virgin Islands, U.S."},
-            {"WF", "Wallis and Futuna"},
-            {"EH", "Western Sahara"},
             {"YE", "Yemen"},
             {"ZM", "Zambia"},
-            {"ZN", "Zimbabwe"}
+            {"ZW", "Zimbabwe"}
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp/Constants/Country.cs:206]: Does changing this country list affect any other functionality not related to this PR? If yes, is it intended?

File: Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakzal) [reply Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs:76]: Looking at the definition, I do not understand what's behind the condition that's supposedly met:

```csharp
 bool conditionMet = propertyValue is List<string> list && list.Contains(containsValue);
```

If any of you understands it and can suggest a better name, I'm all for improving naming and therefore understanding. This whole method could do with some better named variables, but I'd need to spend more time and focus better to suggest anything.

I agree the comment doesn't help and is redundant.

[the remark this answers is not in the document]
Comment (shilpigoel-goaco) [reply Libraries/CO.CDP.Mvc.Validation/RequiredIfAttribute.cs]: renamed to IsDependentMatchWithTarget

File: Frontend/CO.CDP.OrganisationApp/Constants/Country.cs (hunk 7 above)
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Constants/Country.cs:206]: I have updated from the list provided by Sam, and looks like earlier there was mistake which is corrected now.

**Review by SDGoaco — APPROVED** [review]
working as expected in Local Enviroment

File: Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        try
+        {
+            InternationalIdentifiers = await pponClient.GetIdentifierRegistriesAsync(Country);
+        }
+        catch (ApiException ex) when (ex.StatusCode == 404)
+        {
+            InternationalIdentifiers = new List<IdentifierRegistries>();
+        }
+        catch (Exception ex)
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs:70]: Some static analysis warnings reported by github here.

**Review by jakzal — CHANGES_REQUESTED** [review]

**Comment by shilpigoel-goaco** [comment]
> Pro tip for the future - to help maintain a cleaner / leaner history of changes, I'd strongly recommend using the rebase to keep your branch up to date with `main` rather than merging `main` into your branch. With the rebase you don't get merge commits and any conflicts are resolved during rebase, so no extra commits that fix conflicts are required.

**Comment by shilpigoel-goaco** [comment]
> > Pro tip for the future - to help maintain a cleaner / leaner history of changes, I'd strongly recommend using the rebase to keep your branch up to date with `main` rather than merging `main` into your branch. With the rebase you don't get merge commits and any conflicts are resolved during rebase, so no extra commits that fix conflicts are required.

@mareg I do appreciate your tip but I prefer merge into my branch as there are many changes in the main branch frequently and many times same files so resolving conflict becomes very hard with rebase. but definitely something to consider for small changes PR. Thanks

File: Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs (hunk 8 above)
Comment (shilpigoel-goaco) [reply Frontend/CO.CDP.OrganisationApp/Pages/Registration/OrganisationInternationalIdentification.cshtml.cs]: sorted

**Review by jakzal — APPROVED** [review]
